### PR TITLE
feat(wallet): partially persist redux stores

### PR DESCRIPTION
**Pull request by josheleonard**
Resolves https://github.com/brave/brave-browser/issues/29101

Security review: https://github.com/brave/security/issues/1220

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Scenario 1 - Clear storage via wallet reset
1. Create a wallet
2. unlock the wallet
3. open the wallet page
4. refresh the Wallet Page
5. open DevTools
6. Verify that partial application state is saved to local-storage
7. Press ctrl+r (or CMD+R on MacOS) to force-refresh the page
8. Verify that the persisted local-storage state was cleared/reset
10. Repeat steps 1-7, but test the Panel via brave://wallet-panel.top-chrome/ instead of the Page 

Scenario 2 - Clear storage via hard-refresh
1. Create a wallet
2. unlock the wallet
3. open the wallet page
4. refresh the Wallet Page
5. open DevTools
6. Verify that partial application state is saved to local-storage
7. open brave://settings/web3
8. Click "Reset Brave Wallet"
9. Type "Yes" and click OK
11. Open the wallet page
12. Verify that the persisted local-storage state was cleared/reset
13. Repeat steps 2-11, but test the Panel via brave://wallet-panel.top-chrome/ instead of the Page

### CODE REVIEW

File: components/brave_wallet_ui/page/reducers/page_reducer.ts
Comment (diracdeltas) [inline components/brave_wallet_ui/page/reducers/page_reducer.ts:256]: can we be extra sure that the private key (in hex or other form), password, and any auth tokens don't get persisted?

it almost seems like it would be safer to use a whitelist of persistable fields rather than a blacklist. that way if a new auth token or something gets added, there's no chance it would accidentally get saved in plaintext.

File: components/brave_wallet_ui/utils/state-migration-utils.ts
Comment (diracdeltas) [inline components/brave_wallet_ui/utils/state-migration-utils.ts:49]: can persisted API state also be removed when wallet is disabled / deleted / reset?

File: components/brave_wallet_ui/page/reducers/page_reducer.ts
Comment (josheleonard) [reply components/brave_wallet_ui/page/reducers/page_reducer.ts:256]: Agreed. I'll switch these to use whitelists

File: components/brave_wallet_ui/utils/state-migration-utils.ts
Comment (josheleonard) [reply components/brave_wallet_ui/utils/state-migration-utils.ts:49]: I have logged an issue to add the ability to clear this data when the wallet is reset: https://github.com/brave/brave-browser/issues/29138

File: components/brave_wallet_ui/page/wallet_page.tsx
Comment (kdenhartog) [inline components/brave_wallet_ui/page/wallet_page.tsx:38]: nit - non security blocking: We'll probably want to expose this as a button on some pages as well

File: components/brave_wallet_ui/panel/reducers/panel_reducer.ts
Comment (kdenhartog) [inline components/brave_wallet_ui/panel/reducers/panel_reducer.ts:200]: Would it make sense to update this one to an allow list too? What possible panels would be able to be added to this at the moment?

File: components/brave_wallet_ui/common/slices/wallet.slice.ts
Comment (kdenhartog) [inline components/brave_wallet_ui/common/slices/wallet.slice.ts:618]: Same question here, where can I see all the potential ones can be added here? Thinking it would be good to go allow list here at the beginning until we have a better understanding of which data is being persisted

File: components/brave_wallet_ui/common/slices/api.slice.ts
Comment (kdenhartog) [inline components/brave_wallet_ui/common/slices/api.slice.ts:170]: can we make sure to refresh after we see a submitted transaction is completed too?

File: components/brave_wallet_ui/panel/reducers/panel_reducer.ts
Comment (kdenhartog) [reply components/brave_wallet_ui/panel/reducers/panel_reducer.ts:200]: List is available here: https://github.com/brave/brave-core/blob/feat-persist-wallet-redux-store/components/brave_wallet_ui/constants/types.ts#L279

File: components/brave_wallet_ui/common/slices/wallet.slice.ts
Comment (kdenhartog) [reply components/brave_wallet_ui/common/slices/wallet.slice.ts:618]: List is available here: https://github.com/brave/brave-core/blob/feat-persist-wallet-redux-store/components/brave_wallet_ui/constants/types.ts#L229

File: components/brave_wallet_ui/common/slices/api.slice.ts
Comment (josheleonard) [reply components/brave_wallet_ui/common/slices/api.slice.ts:170]: I configured the api slice to refetch any data that is older than 30 seconds old and/or when the app regains focus. Rehydration should only happen on app startup and is used to refresh all of the data within the slice, so it doesn't make sense to me to wipe the entire cache when a transaction updates. Instead, I have updated my PR to ensure that the cached value for a specific transaction will update when the `onTransactionStatusChanged` event is fired from the `wallet_api_proxy`. I need to complete resolve [this issue](https://github.com/brave/brave-browser/issues/28800) in order to properly update the aggregated transaction history

File: package.json
Comment (josheleonard) [reply package.json]: locked: https://github.com/brave/brave-core/pull/17631/commits/37fd447a623d463478def3839ac536f49166b4da

File: package.json
Comment (diracdeltas) [inline package.json:382]: are there any alternative libraries that would work here which are more actively maintained?

File: components/brave_wallet_ui/common/slices/api.slice.ts
Comment (kdenhartog) [reply components/brave_wallet_ui/common/slices/api.slice.ts:170]: Ok, that seems like a better way to handle this case. Good idea!

File: components/brave_wallet_ui/page/reducers/page_reducer.ts
Comment (josheleonard) [reply components/brave_wallet_ui/page/reducers/page_reducer.ts]: Updated to use whitelists

**Comment by josheleonard** [comment]
Noting here that if we want to store the persisted data in a location other than local-storage, a custom interface from core can be utilized as long as it implements `setItem` `getItem` `removeItem` similar to the local-storage api.

File: components/brave_wallet_ui/utils/state-migration-utils.ts
Comment (josheleonard) [reply components/brave_wallet_ui/utils/state-migration-utils.ts:48]: Wallet local storage will now be cleared on wallet reset as of: https://github.com/brave/brave-core/pull/17809

File: components/brave_wallet_ui/common/slices/wallet.slice.ts
Comment (josheleonard) [reply components/brave_wallet_ui/common/slices/wallet.slice.ts]: updated to a whitelist

File: components/brave_wallet_ui/panel/reducers/panel_reducer.ts
Comment (josheleonard) [reply components/brave_wallet_ui/panel/reducers/panel_reducer.ts]: Now using a whitelist

**Review by diracdeltas — APPROVED** [review]
It would be good to add some tests for the following to future-proof this but otherwise lgtm:

1. make sure nothing that looks like a private key or mneumonic is in the serialized persisted data
2. make sure that persisted data is cleared when wallet is reset

File: components/brave_wallet_ui/page/wallet_page.tsx
Comment (kdenhartog) [reply components/brave_wallet_ui/page/wallet_page.tsx:41]: The primary thing I was thinking about here was refresh of portfolio data and NFT data, but I'm not even sure that aligns with this work so no reason to block on it.

**Comment by josheleonard** [comment]
> * make sure nothing that looks like a private key or mneumonic is in the serialized persisted data
> * make sure that persisted data is cleared when wallet is reset

test for "make sure that persisted data is cleared when wallet is reset" should be handled by https://github.com/brave/brave-core/pull/17809/files#diff-c114760829ee3614f7d0bcef91204dc45e57455128412f7b4f66704df7c24bc6R2464-R2487

"make sure nothing that looks like a private key or mnemonic is in the serialized persisted data": I'm thinking of moving the whitelisted keys into a file that requires security approval for changes. That way we wont have to perform checks at runtime

**Comment by josheleonard** [comment]
I moved all of the whitelists definitions to a single file (`components/brave_wallet_ui/common/constants/persisted-state-keys-whitelists.ts`). I will request to Devops to add this file to a list of files that should require security review before merging updates.

**Comment by muliswilliam** [comment]
@kdenhartog Is there anything pending from sec side for this PR?

File: components/brave_wallet_ui/common/constants/persisted-state-keys-whitelists.ts
Comment (kdenhartog) [inline components/brave_wallet_ui/common/constants/persisted-state-keys-whitelists.ts:1]: Thank you for commenting the reason for each one making it onto the blocklists

File: components/brave_wallet_ui/page/wallet_page.tsx
Comment (kdenhartog) [reply components/brave_wallet_ui/page/wallet_page.tsx:40]: fine to ignore this one - I had a different idea about how we might be able to get a user initiated refresh

**Comment by kdenhartog** [comment]
> @kdenhartog Is there anything pending from sec side for this PR?

I think it's mainly discussion around the privacy side of things that ended up holding this one up from looking at the security issue.

When I was just reviewing it now I did become a bit more questionable to what's happening in [brave/brave-core@`e3d4bf7` (#17631)](https://github.com/brave/brave-core/pull/17631/commits/e3d4bf70223f6d122f70b514c7668a5333369fb8). More specifically, it's a bit odd that it looks like we're building an allowlist by taking a list of defined pages and then explicitly excluding a set of keys that are defined via a blocklist. This means that if one of these PageStates were to include an item that's not on the blocklist that we would potentially still cache the item which could present a security concern in the future unless we specifically add it to the blocklist. I think the idea that we wanted to go with here is to build an allowlist by explicitly defining each item here. Am I understanding how this `Exclude<>` logic is working properly?

Main reason I'm asking about this is it looks like this change was made after @diracdeltas dropped the approval on this, so I want to make sure it's aligned with what she had originally suggested in this comment https://github.com/brave/brave-core/pull/17631#discussion_r1139364276

File: components/brave_wallet_ui/common/constants/persisted-state-keys-whitelists.ts
Comment (kdenhartog) [inline components/brave_wallet_ui/common/constants/persisted-state-keys-whitelists.ts:92]: This approach appears to be creating a blocklist here still. Can we make this an allowlist again?

The reason I'm thinking this is the case is because the pageStatePersistorWhitelist is generated by first taking the `PageState` and then excluding the `BlacklistedPageStateKey` from the PageState. Instead, I believe we should be defining this via using `Subset<keyof PageState, {UnionType of explicitly allowed PageStates}>`

File: components/brave_wallet_ui/common/constants/persisted-state-keys-whitelists.ts
Comment (kdenhartog) [inline components/brave_wallet_ui/common/constants/persisted-state-keys-whitelists.ts:167]: Same here for `WalletState` I believe.

File: components/brave_wallet_ui/common/constants/persisted-state-keys-whitelists.ts
Comment (kdenhartog) [reply components/brave_wallet_ui/common/constants/persisted-state-keys-whitelists.ts:92]: Oh wait, are we doing it this way so that the typing system knows to explicitly never accidently include one of the items on the `BlacklistedPageStateKey`? What happens if the `PageState` type contains a value that's not on either `pageStatePersistorWhitelist` or `BlacklistedPageStateKey`?

Comment (josheleonard) [reply components/brave_wallet_ui/common/constants/persisted-state-keys-whitelists.ts:92]: correct, the type system is enforcing that only keys from the state object that have not been blacklisted can be added to the whitelist without throwing a type error. This was to provide a way to provide explicit reasoning as to why properties were whitelisted or blacklisted. The blacklists are types only, and will not be used during runtime

Comment (kdenhartog) [reply components/brave_wallet_ui/common/constants/persisted-state-keys-whitelists.ts]: Cool that makes sense and seems like the best way to leverage the type system. Thanks for clarifying that. I see no new concerns with sec review side of things.

**Review by kdenhartog — APPROVED** [review]

File: components/brave_wallet_ui/page/wallet_page.tsx
Comment (onyb) [inline components/brave_wallet_ui/page/wallet_page.tsx:45]: Discussed internally too, but we probably want to purge for `Shift + Cmd + R`.

File: components/brave_wallet_ui/page/container.tsx
Comment (josheleonard) [reply components/brave_wallet_ui/page/container.tsx]: good catch. fixed

**Comment by josheleonard** [comment]
Closing since partial persistence of the api-slice was not working.
Might re-attempt later without redux-persist and with a way for the api base cache to persist some of it's data to local-storage.

**Pull request closed without merging**
